### PR TITLE
added backtesting to Sao Paulo TRS model notebook

### PR DESCRIPTION
**Pull request by nataliegable-gro**
Description:
Added backtesting for the Sao Paulo TRS model, using the leave-one-year-out cross validation method. The code now includes a function to implement backtesting on all years, report model statistics, and error statistics. There is also a function to plot these model and error statistics over the years used in the leave-one-year-out method. 

Backtesting allows for better model evaluation, as well as giving users/clients an example of the backtesting methods we use in practice.

This notebook also now includes a section on cleaning soil moisture data as a potential input feature (note that soil moisture was not included as an input feature in the final model).

### CODE REVIEW

**Comment by nataliegable-gro** [comment]
<p>:)</p>

---

 <a href='https://app.reviewnb.com/gro-intelligence/api-client/pull/338/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by nataliegable-gro** [comment]
<p>Reformatted wrapped line using PEP 8 style guidelines.</p>

---

 <a href='https://app.reviewnb.com/gro-intelligence/api-client/pull/338/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by nataliegable-gro** [comment]
<p>We remove soil moisture section of this notebook so this issue is resolved.</p>

---

 <a href='https://app.reviewnb.com/gro-intelligence/api-client/pull/338/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by nataliegable-gro** [comment]
<p>Removed SM entirely from notebook.</p>

---

 <a href='https://app.reviewnb.com/gro-intelligence/api-client/pull/338/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by nataliegable-gro** [comment]
<p>Changed the input variable name yield_col to predicted_col.</p>

---

 <a href='https://app.reviewnb.com/gro-intelligence/api-client/pull/338/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by nataliegable-gro** [comment]
<p>Added docstrings for functions get_subharvest_group, run_backtesting_all_subharvest_models, and plot_metric_one_year_left_out.</p>

---

 <a href='https://app.reviewnb.com/gro-intelligence/api-client/pull/338/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by nataliegable-gro** [comment]
<p>Added description of each metric in markdown below this cell. </p>

---

 <a href='https://app.reviewnb.com/gro-intelligence/api-client/pull/338/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by nataliegable-gro** [comment]
<p>Removed this cell and other cells about with feature selection.</p>

---

 <a href='https://app.reviewnb.com/gro-intelligence/api-client/pull/338/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by nataliegable-gro** [comment]
<p>Added note to markdown: We use the input variables associated with the middle subharvest period because we are predicting a date within this period.</p>

---

 <a href='https://app.reviewnb.com/gro-intelligence/api-client/pull/338/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by nataliegable-gro** [comment]
<p>Section on GFS moved up. Section on backtesting moved to the end of the notebook.</p>

---

 <a href='https://app.reviewnb.com/gro-intelligence/api-client/pull/338/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by nataliegable-gro** [comment]
<p>Removed soil moisture from the notebook entirely.</p>

---

 <a href='https://app.reviewnb.com/gro-intelligence/api-client/pull/338/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by nataliegable-gro** [comment]
<p>Thanks! Have changed the function get_subharvest_group to use np.digitize.</p>

---

 <a href='https://app.reviewnb.com/gro-intelligence/api-client/pull/338/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by nataliegable-gro** [comment]
<p>Changed function to return dictionary of lists instead, then using this dictionary directly to populate metric dataframe.</p>

---

 <a href='https://app.reviewnb.com/gro-intelligence/api-client/pull/338/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by nataliegable-gro** [comment]
<p>Looking into this! For client-facing purposes, do we prefer one way or another?</p>

---

 <a href='https://app.reviewnb.com/gro-intelligence/api-client/pull/338/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by jeanzhaoyi** [comment]
<p>We would use <code>year</code>  as a group if we choose to use <code>sklearn.model_selection.LeaveOneGroupOut</code>.  </p><p>Each subharvest_group has its own model.</p><p>I think the current version is a little easier to understand for readers that maybe be unfamiliar with the sklearn methods.</p>

---

 <a href='https://app.reviewnb.com/gro-intelligence/api-client/pull/338/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by nataliegable-gro** [comment]
<p>Updated MAPE description.</p>

---

 <a href='https://app.reviewnb.com/gro-intelligence/api-client/pull/338/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by nataliegable-gro** [comment]
<p>Added comment on backtesting results.</p>

---

 <a href='https://app.reviewnb.com/gro-intelligence/api-client/pull/338/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by nataliegable-gro** [comment]
<p>Ok, great. Leaving the method as is.</p>

---

 <a href='https://app.reviewnb.com/gro-intelligence/api-client/pull/338/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Review by raphael-cohen — APPROVED** [review]
Looks good to me
